source/mod_boinc.f90

```diff
@@ -0,0 +1,196 @@
+module mod_boinc
+
+  use floatPrecision
+
+  implicit none
+
+  integer,            private, save :: boinc_nTurn        = 0
+  integer,            private, save :: boinc_cpInterval   = -1
+  integer,            private, save :: boinc_progInterval = -1
+  integer,            private, save :: boinc_logUnit      = -1
+  character(len=12),  private, save :: boinc_logFile      = "cr_boinc.log"
+  character(len=256), private, save :: boinc_logBuffer    = " "
+  logical,            private, save :: boinc_isStandalone = .false.
+  real(kind=fPrec),   private, save :: boinc_lastCPReq    = 0.0
+  real(kind=fPrec),   private, save :: boinc_lastProgress = 0.0
+
+contains
+
+! ================================================================================================ !
+!  Open the BOINC log file
+!  Initialise BOINC
+!  Set the checpointing intervals depending on whether BOINC is running in standalone mode or not
+! ================================================================================================ !
+subroutine boinc_initialise
+
+  use mod_units
+  use mod_common
+
+  integer tmpInt
+
+  call f_requestUnit(boinc_logFile, boinc_logUnit)
+  call f_open(unit=boinc_logUnit,file=boinc_logFile,formatted=.true.,mode="w+")
+  write(boinc_logBuffer,"(a)") "Initialising BOINC"
+  call boinc_writeLog
+
+#ifdef API
+  call boinc_init
+  call boinc_is_standalone(tmpInt)
+  boinc_isStandalone = (tmpInt /= 0)
+#else
+  boinc_isStandalone = .true.
+  write(boinc_logBuffer,"(a)") "BOINC API not available"
+  call boinc_writeLog
+#endif
+  write(boinc_logBuffer,"(a,l1)") "Standalone: ",boinc_isStandalone
+  call boinc_writeLog
+
+  if(boinc_isStandalone) then
+    boinc_cpInterval   = 10
+    boinc_progInterval = 1
+  else
+    boinc_cpInterval   = 120
+    boinc_progInterval = 1
+  end if
+
+end subroutine boinc_initialise
+
+! ================================================================================================ !
+!  Called on every turn in tracking
+!  Report progress to BOINC maximum every boinc_progInterval second
+!  Checkpoint every boinc_cpInterval seconds, but only if the API allows it
+! ================================================================================================ !
+subroutine boinc_turn(nTurn)
+
+  use mod_common
+  use checkpoint_restart
+
+  integer, intent(in) :: nTurn
+
+  integer          doCheckpoint
+  real(kind=fPrec) cpuTime, boincProg
+
+  boinc_nTurn = nTurn
+  call cpu_time(cpuTime)
+
+#ifdef API
+  if(cpuTime-boinc_lastProgress >= boinc_progInterval) then
+    ! Tell BOINC how we're doing
+    call boinc_fraction_done(dble(nTurn)/dble(numl))
+    boinc_lastProgress = cpuTime
+  end if
+#endif
+
+  if(cpuTime-boinc_lastCPReq < boinc_cpInterval .and. nTurn > 1) then
+    return ! Not checkpointing yet, just return
+  end if
+
+  write(boinc_logBuffer,"(a,f0.2,a)") "Last checkpoint ",cpuTime-boinc_lastCPReq," seconds ago"
+  call boinc_writeLog
+
+  if(cr_checkp) then
```

Comment:
is this `if` statement necessary if we have the BOINC API?
I think that the BOINC API implies that we have checkpoint-restart...

Reply:
I left it in because it was there in the old BOINC/CR routine, but after having had a look in the source code.

It seems to me that it was used to prevent any of the C/R routines from being called in the DA version of SixTrack. The flag is only set true in main_cr. The CR core code is still built into the DA version, but none of the routines that call C/R are built in any more. I think we can safely remove it.

Reply:
Removed it, and it runs fine.